### PR TITLE
lwt: Cassandrda compatibility for treatment of static rows

### PR DESCRIPTION
**Pull request by kostja**
lwt: Cassandrda compatibility when incarnating a row for UPDATE
When evaluating an LWT condition involving both static and non-static
cells, and matching no regular row, the static row must be used UNLESS
the IF condition is IF EXISTS/IF NOT EXISTS, in which case special rules
apply.

Before this fix, Scylla used to assume a row doesn't exist if there is
no matching primary key. In Cassandra, if there is a
non-empty static row in the partition, a regular row based
on the static row' cell values is created in this case, and then this
row is used to evaluate the condition.

This problem was reported as https://github.com/scylladb/scylla/issues/10081.

The reason for Scylla behaviour before the patch was that when
implementing LWT I tried to converge Cassandra data model (or lack of
thereof) with a relational data model, and assumed a static row is a
"shared" portion of a regular row, i.e. a storage level concept intended
to save space, and doesn't have independent existence.
This was an oversimplification.

This patch fixes https://github.com/scylladb/scylla/issues/10081, making Scylla semantics match the one of
Cassandra.

I will now list other known examples when a static row has an own
independent existence as part of a table, for cataloguing purposes.

SELECT * from a partition which has a partition key
and a static cell set returns 1 row. If later a regular row is added
to the partition, the SELECT would still return 1 row, i.e.
the static row will disappear, and a regular row will appear instead.

Another example showing a static row has an independent existence below:

CREATE TABLE t (p int, c int, s int static, PRIMARY KEY(p, c));
INSERT INTO t (p, c) VALUES(1, 1);
INSERT INTO t (p, s) VALUES(1, 1) IF NOT EXISTS;

In Cassandra (and Scylla), IF NOT EXISTS evaluates to TRUE, even though both
the regular row and the partition exist. But the static cells are not
set, and the insert only provides a partition key, so the database assumes the
insert is operating against a static row.

It would be wrong to assume that a static row exists when the partition
key exists:
INSERT INTO t (p, c, s) VALUES(1, 1, 1) IF NOT EXISTS;

 [applied] | p | c | s
 -----------+---+---+------
      False | 1 | 1 | null

evaluates to False, i.e. the regular row does exist when p and c exist.

Issue

CREATE TABLE t (p INT, c INT, r INT, s INT static, PRIMARY KEY(p, c))
INSERT INTO t (p, s) VALUES (1, 1);
UPDATE t SET s=2, r=1 WHERE p=1 AND c=1 IF s=1 and r=null;
- in this case, even though the regular row doesn't exist, the static
row does, and should be used for condition evaluation.

In other words, IF EXISTS/IF NOT EXISTS have contextual semantics.
They apply to the regular row if clustering key is used in the WHERE
clause, otherwise they apply to static row.

One analogy for static rows is that it is like a static member of C++ or
Java class. It's an attribute of the class (assuming class = partition),
which is accessible through every object of the class (object = regular
row). It is also present if there are no objects of the class, but the
class itself exists: i.e. a partition could have no regular rows, but
some static cells set, in this case it has a static row.

*Unlike C++/Java static class members* a static row is an optional
attribute of the partition. A partition may exist, but the static row
may be absent (e.g. no static cell is set). If the static row does exist,
all regular rows share its contents, *even if they do not exist*.
A regular row exists when its clustering key is present
in the table. A static row exists when at least one static cell is set.

Tests are updated because now when no matching row is found
for the update we show the value of the static row as the previous
value, instead of a non-matching clustering row.

### CODE REVIEW

**Comment by nyh** [comment]
@kostja please check https://jenkins.scylladb.com/job/releng/job/Scylla-CI/653/ why the CI run failed. I am guessing it is some problem not related at all to your patch, but I want you to take a look at the Jenkins output yourself - I want you to experience first-hand the difficulty of understanding which test failed and why, maybe then you'll understand my dislike of 2-line test reports that send you to other files, instead of test reports that include the content of those files.

**Comment by kostja** [comment]
@nyh I will address comment 1, re "fixes". Comment 2 and 3 I am not going to address since I disagree with them. I don't know why you created this test, I find it useless. As well as I find it useless to compare with Cassandra in this context. We can discuss it more, if you like, but I don't know what kind of argument should we come up with since I feel pretty strongly about how this should be tested. This pull request addresses Gleb's review request to make it clearly visible what cells are now reported as null cells. With tabular output that cql tests switched to this is now clearly visible. And, last note, I find it *extremely* unproductive to argue about some "philosophical" issues as you put it, while there is a community user waiting for a fix. There are no comments on the patch itself, so why is it being held up?

As to failing CI, yes, I am looking into it, but it's a new flaky test - next promotion fails on the same issue. I posted the root cause in #raft channel.

**Comment by kostja** [comment]
Re experience first-hand - I got in touch with @benipeled already to make sure the logs from individual tests are captured in testlog. The relevant part is also captured in test output which allowed me to analyze this issue and connect it with next promotion failures.

**Comment by avikivity** [comment]
> I must begin by admitting that I lack mental capacity for fully understanding what Cassandra static rows are.
> 
> My current mental model is as follows:
> 
> A static row is like a static member of C++ or Java class. It's an attribute of the class (assuming class = partition), which is accessible through every object of the class (object = regular row). It is also present if there are no objects of the class, but the class itself exists: i.e. a partition could have no regular rows, but some static cells set, in this case it has a static row.
> 
> _Unlike C++/Java static class members_ a static row is an optional attribute of the partition. A partition may exist, but the static row may be absent (e.g. no static cell is set). If the static row does exist, all regular rows share its contents, _even if they do not exist_. A regular row exists when its clustering key is present in the table. A static row exists when at least one static cell is set.
> 


My model is a bit different:
- if regular rows exist, then they are extended with the contents of the static row. e.g. after inserting one regular row and one static row into an empty partition, a SELECT returns one row.
- if regular rows do not exist, then the static row appears as an individual row. After inserting just static columns into an empty partition, a SELECT returns one row.

**Comment by nyh** [comment]
@kostja I don't know how you can title your patch "Cassandra compatibility" and at the same time insist that you "find it useless to compare with Cassandra in this context".
Apparently I'm not going to convince you. I'll write the compatibility test myself after your patch goes in (and as noted, I even wrote a big part of it already and it's marked xfail).

**Comment by kostja** [comment]
@nyh I explained it on the mailing list, let me explain again. *Unlike* Cassandra, Scylla returns a row for *every* statement of the batch, in order of statements in the batch. In other words, we're incompatible by design here. This is what the changes in the existing tests highlight - they highlight the changes in *our own* implementation. What the fix is doing it is changing *what* rows are returned. Previously, for every missing row we would return the first row in the partition. Now, we return a new "incarnated" row, which is based on the static row, but has nulls for all non-static columns. In this, we're becoming compatible with Cassandra. 

Writing a pytest *just* to capture this part of compatibility would be of little value since we will need a CQL test that would capture the output as well, and such test would capture the compatibility part too, and such test is already part of the patch.

**Comment by nyh** [comment]
I am not convinced. As I keep saying, you titled the patch "improve Cassandra compatibility", but you don't have a test that can back up that specific claim because nothing you can do by testing only on Scylla can convince me about Cassandra compatibility. 

You say that Scylla is deliberately different from Cassandra in one specific thing, but then 1. We never documented this difference in our cql differences document, and 2. We never tried to prove with a test that this deliberate difference is the *only* difference. And 3. A test can be written to check the fix in this issue while ignoring the by-design difference - this is one of the benefits of test-with-python-code compared to test-by-literal-comparison-of-output.

In fact, this compatibility bug could exist until now exactly because we didn't compare Scylla and Cassandra in this static-row scenario. This comparison is exactly what my xfailing test did, and it demonstrates the bug which you fixed in this patch in fact existed. I can't for the life of me understand why you "don't like" a test which reproduced the specific bug which you claim to have fixed - and why you didn't even check if your patch made it pass (by the way, if it does pass now, please remove the xfail tag. If it doesn't pass, I wonder why?).

As I said, I have no plan or desire to force you to write such a compatibility test. I just claimed it is necessary and you did not convince me otherwise. But you don't have to write it yourself - I can, and I will. So this doesn't have to delay committing your patch.

**Comment by kostja** [comment]
I updated the commit.

File: cql3/statements/cas_request.cc [hunk 1]
```diff
@@ -120,21 +120,23 @@ std::optional<mutation> cas_request::apply(foreign_ptr<lw_shared_ptr<query::resu
...
-    const clustering_key& ckey = !op.statement.has_only_static_column_conditions() && op.ranges.front().start() ?
-        op.ranges.front().start()->value() : empty_ckey;
-    return _rows.find_row(pkey, ckey);
+    const clustering_key& ckey = op.ranges.front().start() ?  op.ranges.front().start()->value() : empty_ckey;
+    auto row = _rows.find_row(pkey, ckey);
+    if (row == nullptr && !ckey.is_empty() &&
+        !op.statement.has_if_exist_condition() && !op.statement.has_if_not_exist_condition()) {
+        row = _rows.find_row(pkey, empty_ckey);
+    }
+    return row;
```
Comment (avikivity) [inline cql3/statements/cas_request.cc:139]: I don't know if this is correct. I think the static-columns-only row materializes only if there are no rows at all.

File: test/cql/cassandra_cql_test.result [hunk 2]
```diff
@@ -755,12 +755,12 @@ OK
 > INSERT INTO lwt_on_static_columns_with_null_batch (a, b, d) values (2, 2, 'a')
 > UPDATE lwt_on_static_columns_with_null_batch  SET s = 2 WHERE a = 2 IF s = null
```
Comment (avikivity) [inline test/cql/cassandra_cql_test.result:756]: Wow, this `s = NULL` is sad. Won't it return NULL unconditionally in SQL?

Comment (avikivity) [reply test/cql/cassandra_cql_test.result:756]: (and `IS NULL` needs to be used to detect NULLs?)

**Comment by kostja** [comment]
* Avi Kivity ***@***.***> [22/06/13 16:17]:
> 
> >      // turn into INSERT), because, although the table doesn't have any regular rows matching the
>      // statement clustering column restriction, the static row matches the statement condition:
>      //   CREATE TABLE t(p int, c int, s int static, v int, PRIMARY KEY(p, c));
>      //   INSERT INTO t(p, s) VALUES(1, 1);
>      //   UPDATE t SET v=1 WHERE p=1 AND c=1 IF s=1;
> -    // Another case when we pass an empty clustering key prefix is apparently when the table
> -    // doesn't have any clustering key columns and the clustering key range is empty (open
> -    // ended on both sides).
> -    const clustering_key& ckey = !op.statement.has_only_static_column_conditions() && op.ranges.front().start() ?
> -        op.ranges.front().start()->value() : empty_ckey;
> -    return _rows.find_row(pkey, ckey);
> +    const clustering_key& ckey = op.ranges.front().start() ?  op.ranges.front().start()->value() : empty_ckey;
> +    auto row = _rows.find_row(pkey, ckey);
> +    if (row == nullptr && !ckey.is_empty() &&
> +        !op.statement.has_if_exist_condition() && !op.statement.has_if_not_exist_condition()) {
> +        row = _rows.find_row(pkey, empty_ckey);
> +    }
> +    return row;

> 
> I don't know if this is correct. I think the static-columns-only
> row materializes only if there are no rows at all.

The test case given by the user is about exactly this, and 
and there is another row in the partition already.
https://github.com/scylladb/scylla/issues/10081

But it's a good point that *when* to materialize a static row is
also a decision point. 

-- 
Konstantin Osipov, Moscow, Russia

**Comment by kostja** [comment]
* Avi Kivity ***@***.***> [22/06/13 16:17]:
> > > My model is a bit different:
> > > 
> > > * if regular rows exist, then they are extended with the contents of the static row. e.g. after inserting one regular row and one static row into an empty partition, a SELECT returns one row.
> > > * if regular rows do not exist, then the static row appears as an individual row. After inserting just static columns into an empty partition, a SELECT returns one row.
> > 
> > Avi, while your observation of what happens is correct, it is not a valid data model. In relational algebra, UPDATE and SELECT are operators which are applied to a relation. You can't have a row exist in UPDATE and not EXIST in SELECT - because they work with the same result set. Similarly, you can't have a row exist in SELECT but then disappear once you insert another row. Imagine a bag of items. You can't have an item disappear from the bag depending on whether you're adding a thing into the bag, or taking a thing from the bag.
> 
> Agree, but what are you suggesting? That we change the data model? That can break applications.

I suggest we try to explain things in terms of a data model and
representation rules. We should define existence rules for a
static row independently on search rules, and explain that
sometimes we hide the static row in the output because of <silly
reasons>.

In the comment I am trying to come up with some rules that explain
when a static row comes into existence and what it belongs to 
(shared by regular rows, part of the partition, which becomes an
independent entity, etc). 

With this commit, I am abandoning the idea that Cassandra data
model is compatible with a relational one.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by kostja** [comment]
What exactly needs to be done with this patch now? Let's queue it already!

File: cql3/statements/cas_request.cc (hunk 1 above)
Comment (kostja) [reply cql3/statements/cas_request.cc:139]: The test case given by the user is about exactly this, and
and there is another row in the partition already.
[#10081](https://github.com/scylladb/scylla/issues/10081)

But it's a good point that *when* to materialize a static row is
also a decision point.

File: test/cql/cassandra_cql_test.result (hunk 2 above)
Comment (kostja) [reply test/cql/cassandra_cql_test.result:756]: s = NULL would return a syntax error in SQL. But it depends on the dialect. IS NULL needs to be used to detect NULLs.

**Comment by avikivity** [comment]
> * Avi Kivity ***@***.***> [22/06/14 10:03]:
> @avikivity commented on this pull request. > @@ -120,21 +120,23 @@ std::optional<mutation> cas_request::apply(foreign_ptr<lw_shared_ptr<query::resu const update_parameters::prefetch_data::row* cas_request::find_old_row(const cas_row_update& op) const { static const clustering_key empty_ckey = clustering_key::make_empty(); const partition_key& pkey = _key.front().start()->value().key().value(); - // If a statement has only static columns conditions, we must ignore its clustering columns - // restriction when choosing a row to check the conditions, i.e. choose any partition row, - // because any of them must have static columns and that's all we need to know if the - // statement applies. For example, the following update must successfully apply (effectively + // We must ignore statement clustering column restriction when + // choosing a row to check the conditions. If there is no I don't understand this comment. Why do we have to ignore the clustering key? Maybe you meant "regular column" here?
> This comment is for the following case: - the WHERE clause restricts the partition key and a non-existent clustering key - there is more than one row in the table - there is no match.
> Let's consider some examples. `WHERE pk = :pk AND ck = :ck IF regular_col = :regular AND static_col = :static` Here, we must read the existing clustering row and check if regular_col is correct.
> A parameter marker can be null so I don't see how the examples are different, but ok. The procedure is not exactly like you described. If the regular row exists, yes, then it should be used. If it doesn't exist, we must return the static row.
> `WHERE pk = :pk AND ck = :ck IF regular_col IS NULL AND static_col = :static` Here, we must read the clustering row and allow the update if the row did not exist, or if regular_col was NUKK.
> Scylla doesn't have IS NULL, it has regular_col = null. Then again, not exactly, we only allow the update iff IF condition matches. So we must generate an interim row to evaluate it. So here we also return the static row if there is no match.

I guess this is due to update=insert in CQL. So WHERE ... in an UPDATE has to pretend the row exists. Very confusing. Update=insert was not mandatory in lwt, but if that's how Cassandra implements it we should follow.

btw, and unrelated, we should move away from = NULL and towards IS NULL, introducing a flag to switch dialects. Very sad.


> `WHERE pk = :pk AND ck = :ck IF static_col = :static` Here we don't need to read that clustering row, only the static row.
> And here we should return the static row as well, yes.
> […](#)
> -- Konstantin Osipov, Moscow, Russia

**Pull request closed without merging**